### PR TITLE
Apply PEC to split pml fields

### PR DESCRIPTION
**Pull request by RevathiJambunathan**
This PR fixes a bug that is observed when pml boundary is adjacent to the pec boundary.

The input to reproduce this error : 
[inputs_pec_pml.txt](https://github.com/ECP-WarpX/WarpX/files/7514589/inputs_pec_pml.txt)

The absorption is not clean at the pml boundary (at zmax boundary in this case) when the adjacent boundaries are pec (ymin, and ymax boundary in this case)
dev

Extending the pec boundary into the pml region by applying pec to the splits fields fixes this issue
fix

### CODE REVIEW

File: Source/BoundaryConditions/WarpX_PEC.cpp [hunk 1]
```diff
@@ -49,22 +49,32 @@ PEC::ApplyPECtoEfield (std::array<std::unique_ptr<amrex::MultiFab>, 3>& Efield,
...
-    int nComp_x = Efield[0]->nComp();
-    int nComp_y = Efield[1]->nComp();
-    int nComp_z = Efield[2]->nComp();
+    // For each Efield multifab, apply PEC boundary condition to ncomponents
+    // If not split E-field, the PEC is applied to the regular Efield used in Maxwell's eq.
+    // If split_pml_field is true, then PEC is applied to the first two components of the pml multifabs
+    // Since PML_components are xy, xz for Efield[0]
+    //                          yz, yx for Efield[1]
+    //                          zx, zy for Efield[2]
+    int nComp_x = (split_pml_field) ? 2 : Efield[0]->nComp();
```
Comment (RevathiJambunathan) [inline Source/BoundaryConditions/WarpX_PEC.cpp]: @RemiLehe  @dpgrote Please let me know if you have a suggestion here to avoid using 2 as a hard-coded integer and use ncomp from split fields. Unless divE cleaning is on, pml multifabs do not have 3 components, they always have 2. But I wanted to keep this general. So let me know if you have a more elegant way of doing this

File: Source/BoundaryConditions/WarpX_PEC.cpp [hunk 2]
```diff
@@ -49,22 +49,32 @@ PEC::ApplyPECtoEfield (std::array<std::unique_ptr<amrex::MultiFab>, 3>& Efield,
...
+    int nComp_y = (split_pml_field) ? 2 : Efield[1]->nComp();
+    int nComp_z = (split_pml_field) ? 2 : Efield[2]->nComp();
 #ifdef AMREX_USE_OMP
 #pragma omp parallel if (amrex::Gpu::notInLaunchRegion())
 #endif
-    for (amrex::MFIter mfi(*Efield[0], amrex::TilingIfNotGPU()); mfi.isValid(); ++mfi) {
+    for (amrex::MFIter mfi(*Efield[0], false); mfi.isValid(); ++mfi) {
```
Comment (RevathiJambunathan) [inline Source/BoundaryConditions/WarpX_PEC.cpp]:
```suggestion
    for (amrex::MFIter mfi(*Efield[0], amrex::TilingIfNotGPU()); mfi.isValid(); ++mfi) {
```

**Review by RevathiJambunathan — COMMENTED** [review]
a self-review

File: Source/BoundaryConditions/WarpX_PEC.cpp (hunk 1 above)
Comment (EZoni) [reply Source/BoundaryConditions/WarpX_PEC.cpp]: Good point! I'm thinking, why don't we simply call the method `nComp()` regardless of whether it's called from a MultiFab in the valid domain or from the corresponding MultiFab in the PML (so simply `int nComp_x = Efield[0]->nComp();` and so on, as it was before)? I think such a call would return `1` component for the MultiFabs in the valid domain and `2` or `3` components, automatically, for the MultiFabs in the PML (`2` if div(E) cleaning is OFF, `3` if div(E) cleaning is ON). Do you think this would work or am I overlooking something?

Another separate question (for @RevathiJambunathan, @RemiLehe and @dpgrote): Do we need to apply these boundary conditions to the F MultiFab as well, when div(E) cleaning is ON, both in the valid domain (where F is a scalar) and in the PML (where F has three components Fx, Fy, Fz)?

File: Source/BoundaryConditions/WarpX_PEC.H [hunk 3]
```diff
@@ -276,12 +276,15 @@ using namespace amrex;
...
+     * \param[in]     lev             level of the Multifab
+     * \param[in]     patch_type      coarse or fine
+     * \param[in]     split_pml_field whether pml the multifab is the regular Efield or
+     *                                split pml field
      */
-    void ApplyPECtoEfield ( std::array<std::unique_ptr<amrex::MultiFab>, 3>& Efield,
-                            const int lev, PatchType patch_type);
+    void ApplyPECtoEfield ( std::array<amrex::MultiFab*, 3> Efield,
+                            const int lev, PatchType patch_type,
+                            bool split_pml_field = false);
```
Comment (EZoni) [inline Source/BoundaryConditions/WarpX_PEC.H]:
```suggestion
                            const bool split_pml_field = false);
```

File: Source/BoundaryConditions/WarpX_PEC.cpp [hunk 4]
```diff
@@ -62,9 +68,13 @@ PEC::ApplyPECtoEfield (std::array<std::unique_ptr<amrex::MultiFab>, 3>& Efield,
...
-        amrex::Box const& tex = mfi.tilebox(Efield[0]->ixType().toIntVect(), shape_factor);
-        amrex::Box const& tey = mfi.tilebox(Efield[1]->ixType().toIntVect(), shape_factor);
-        amrex::Box const& tez = mfi.tilebox(Efield[2]->ixType().toIntVect(), shape_factor);
+        // if split field, the box includes nodal flag, else nodal flag plus particle shape factor
+        amrex::Box const& tex = (split_pml_field) ? mfi.tilebox(Efield[0]->ixType().toIntVect())
+                                                  : mfi.tilebox(Efield[0]->ixType().toIntVect(), shape_factor);
+        amrex::Box const& tey = (split_pml_field) ? mfi.tilebox(Efield[1]->ixType().toIntVect())
+                                                  : mfi.tilebox(Efield[1]->ixType().toIntVect(), shape_factor);
+        amrex::Box const& tez = (split_pml_field) ? mfi.tilebox(Efield[2]->ixType().toIntVect())
+                                                  : mfi.tilebox(Efield[2]->ixType().toIntVect(), shape_factor);
```
Comment (EZoni) [inline Source/BoundaryConditions/WarpX_PEC.cpp]: Could you remind of what is the effect of passing the additional argument `shape_factor` to the `tilebox` method here?

File: Source/BoundaryConditions/WarpX_PEC.cpp [hunk 5]
```diff
@@ -49,9 +49,15 @@ PEC::ApplyPECtoEfield (std::array<std::unique_ptr<amrex::MultiFab>, 3>& Efield,
     amrex::IntVect Ex_nodal = Efield[0]->ixType().toIntVect();
     amrex::IntVect Ey_nodal = Efield[1]->ixType().toIntVect();
     amrex::IntVect Ez_nodal = Efield[2]->ixType().toIntVect();
-    int nComp_x = Efield[0]->nComp();
-    int nComp_y = Efield[1]->nComp();
-    int nComp_z = Efield[2]->nComp();
+    // For each Efield multifab, apply PEC boundary condition to ncomponents
+    // If not split E-field, the PEC is applied to the regular Efield used in Maxwell's eq.
+    // If split_pml_field is true, then PEC is applied to the first two components of the pml multifabs
```
Comment (EZoni) [inline Source/BoundaryConditions/WarpX_PEC.cpp]: In this comment I would simply write "PEC is applied to all components of the PML MultiFabs`. I mean, I think we want to apply it to all components, rather than just the first two, or am I misunderstanding something?

File: Source/BoundaryConditions/WarpXFieldBoundaries.cpp [hunk 6]
```diff
@@ -36,9 +51,13 @@ void WarpX::ApplyBfieldBoundary (const int lev, PatchType patch_type, DtType a_d
...
-            PEC::ApplyPECtoBfield( Bfield_fp[lev], lev, patch_type);
+            PEC::ApplyPECtoBfield( { get_pointer_Bfield_fp(lev, 0),
+                                     get_pointer_Bfield_fp(lev, 1),
+                                     get_pointer_Bfield_fp(lev, 2) }, lev, patch_type);
         } else {
-            PEC::ApplyPECtoBfield( Bfield_cp[lev], lev, patch_type);
+            PEC::ApplyPECtoBfield( { get_pointer_Bfield_cp(lev, 0),
+                                     get_pointer_Bfield_cp(lev, 1),
+                                     get_pointer_Bfield_cp(lev, 2) }, lev, patch_type);
         }
```
Comment (EZoni) [inline Source/BoundaryConditions/WarpXFieldBoundaries.cpp:61]: I guess we will want to implement similar changes here for PMC in case we have split components for B in the PML, correct? Would make sense to do this in a separate PR of course, just making a note here.

**Review by EZoni — COMMENTED** [review]
Thank you for this fix, Reva! I left only a few questions or comments.

File: Source/BoundaryConditions/WarpXFieldBoundaries.cpp (hunk 6 above)
Comment (RevathiJambunathan) [reply Source/BoundaryConditions/WarpXFieldBoundaries.cpp:61]: yes - and PMC boundary condition is not supported. Will add this to the Boundary condition project.

File: Source/BoundaryConditions/WarpX_PEC.cpp (hunk 5 above)
Comment (RevathiJambunathan) [reply Source/BoundaryConditions/WarpX_PEC.cpp]: the third one is the normal component (I think?) @RemiLehe @dpgrote may know more. We apply PEC only to the tangential components of the E-field. The third component when divE cleaning is on, is xx (definition in PMLComponent.H). We do not apply PEC to that component.

File: Source/BoundaryConditions/WarpX_PEC.cpp (hunk 1 above)
Comment (RevathiJambunathan) [reply Source/BoundaryConditions/WarpX_PEC.cpp]: See comment above. The reason I asked is because, when nComp() is 3 (when divE cleaning is on), the 3rd component is not tangential to the surface, but normal . Atleast that is what I understand from `xx` defined in PMLComponent.H. And PEC should only set the tangential components of the E-field to 0.

File: Source/BoundaryConditions/WarpX_PEC.cpp (hunk 4 above)
Comment (RevathiJambunathan) [reply Source/BoundaryConditions/WarpX_PEC.cpp]: for the Efields used in the valid region, when we pass the shape factor, it will "grow" the box into the guard cell region by the number of indices equal to the shape factor. We need to include this in addition to just the nodality because particles gather fields from these cells as well. And the guard cells are treated separately in the PEC module.

File: Source/BoundaryConditions/WarpX_PEC.cpp (hunk 1 above)
Comment (RevathiJambunathan) [reply Source/BoundaryConditions/WarpX_PEC.cpp]: I am not familiar with the F Multifab.? PEC is only applied to the electric fields. But I understand that you ask this specifically because of divE cleaning. We do not apply PEC in the valid region to F multifab when divE cleaning is on either. so I dont think we need it here. If we do, then the same thing will need to be implemented in the valid region (at the boundary) as well, I would think

File: Source/BoundaryConditions/WarpX_PEC.cpp (hunk 4 above)
Comment (EZoni) [reply Source/BoundaryConditions/WarpX_PEC.cpp]: Thank you for the clarification, Reva! I wonder then if it would be safer to grow this by a number of guard cells equal to the `IntVect` `ng_FieldGather` that we have in Source/Parallelization/GuardCellManager.H (which would probably require to defined some "getter" similar to `getngE()` and others that already exist in the `WarpX` class). What do you think? Would probably be worth looking briefly at how `ng_FieldGather` is computed in the `guardCellManager` class, it is probably equal to the shape factor by default, but might be taking into account some other algorithmic variants as well.

File: Source/BoundaryConditions/WarpX_PEC.cpp (hunk 1 above)
Comment (EZoni) [reply Source/BoundaryConditions/WarpX_PEC.cpp]: Thank you for the clarifications, Reva! I added a point on the agenda for tomorrow's meeting to discuss what to do with F (and, if necessary, what to do with the diagonal components of E, such `xx` etc.).

Comment (RevathiJambunathan) [reply Source/BoundaryConditions/WarpX_PEC.cpp]: great! I was just about to -- you beat me to it!!!! :)

**Comment by RevathiJambunathan** [comment]
Thanks for meeting to discuss this @EZoni @RemiLehe @jlvay @ax3l @lgiacome 
Some notes here : 
1. PEC will be applied to all split-fields of the tangential E-field component
2. PEC should be applied to the F-fields when divE cleaning is on (this will be a separate PR)
3. Extend the mirroring of the Efields into the guard cell using ng_FieldGather since that IntVect stores the number of cells the particles gather fields from in the guard cell. (and it could be more than just the shape factor)
4. Ideally, when PEC is applied to the pml multifabs, for continuity, we need to apply the mirroring of fields across the PEC boundary. But, this will be addressed later, possibly after incorporating one pml implementation (inside domain versus outside domain)

To complete this PR, (1) and (3) will be added.
(2) is a separate PR
(4) Will be re-visited after finalizing pml location (in/outsite the domain)

**Comment by RevathiJambunathan** [comment]
I have added (1) and (3) in this commit.
5d61e17 
and re-did the test in the description and still looks good!

File: Source/BoundaryConditions/WarpX_PEC.cpp [hunk 7]
```diff
@@ -154,9 +148,9 @@ PEC::ApplyPECtoBfield (std::array<amrex::MultiFab*, 3> Bfield, const int lev,
         amrex::Array4<amrex::Real> const& Bz = Bfield[2]->array(mfi);
 
         // Extract tileboxes for which to loop
```
Comment (EZoni) [inline Source/BoundaryConditions/WarpX_PEC.cpp:150]: Would you mind copying here the same comment that you added for the case of the E field above? I think it's a good explanatory comment, definitely worth having it here as well.

**Review by EZoni — COMMENTED** [review]
Thank you for making the latest changes, Reva! The PR looks good to me. I just left a minor request about a comment that would be nice to duplicate for the parts of the code relevant for both E and B.

File: Source/BoundaryConditions/WarpX_PEC.H (hunk 3 above)
Comment (RevathiJambunathan) [reply Source/BoundaryConditions/WarpX_PEC.H]: I have added this change in the .H and corresponding .cpp file. Thank you so much for this suggestion

File: Source/BoundaryConditions/WarpX_PEC.cpp (hunk 7 above)
Comment (RevathiJambunathan) [reply Source/BoundaryConditions/WarpX_PEC.cpp:150]: Thank you for the suggestion. Added comment here. 6e884c7 
Also include the `const` suggestion in the .H and .cpp file for the ApplyPECtoEfield function argument.

File: Source/BoundaryConditions/WarpX_PEC.cpp (hunk 4 above)
Comment (RevathiJambunathan) [reply Source/BoundaryConditions/WarpX_PEC.cpp]: Thank you Edoardo. we use ng_FieldGather now. :)

File: Source/BoundaryConditions/WarpX_PEC.cpp (hunk 5 above)
Comment (RevathiJambunathan) [reply Source/BoundaryConditions/WarpX_PEC.cpp]: As discussed in an offline meeting, PEC is applied to all split fields of the tangential components

**Review by EZoni — APPROVED** [review]
Thank you for the bug fix and for the changes, Reva! Looks good to me.
